### ISSUE TITLE
[QAT][quant] QAT training can't converge

### ISSUE THREAD

**Issue by JH-Xie** [pytorch/pytorch#79862]
### 🐛 Describe the bug

I use a mobilenetv2 from timm respository, and modify:
1. relu6 -> relu
2. '+' -> nn.quantized.FloatFunctional().add
3. add quant and dequant

then I use a small learning rate to fintune a model pretrained in float32, but the loss can't converge

```
2022-06-20 07:39:54 - INFO - Epoch: [0] [ 0/10640] lr:0.0000001object_loss:1.0642 object_precision: 73.17 object_recall: 70.59 object_f1: 71.86

2022-06-20 07:40:39 - INFO - Epoch: [0] [ 100/10640] lr:0.0000001object_loss:3.7584 object_precision: 47.04 object_recall: 25.56 object_f1: 32.71

2022-06-20 07:41:20 - INFO - Epoch: [0] [ 200/10640] lr:0.0000001 object_loss:3.9367 object_precision: 43.73 object_recall: 23.27 object_f1: 30.02

2022-06-20 07:42:02 - INFO - Epoch: [0] [ 300/10640] lr:0.0000001 object_loss:3.9601 object_precision: 43.11 object_recall: 22.70 object_f1: 29.43

2022-06-20 07:42:43 - INFO - Epoch: [0] [ 400/10640] lr:0.0000001 object_loss:3.9897 object_precision: 43.31 object_recall: 22.65 object_f1: 29.43

2022-06-20 07:43:25 - INFO - Epoch: [0] [ 500/10640] lr:0.0000001 object_loss:3.9930 object_precision: 42.84 object_recall: 22.39 object_f1: 29.10

2022-06-20 07:44:06 - INFO - Epoch: [0] [ 600/10640] lr:0.0000001 object_loss:4.0291 object_precision: 42.62 object_recall: 22.07 object_f1: 28.78

2022-06-20 07:44:47 - INFO - Epoch: [0] [ 700/10640] lr:0.0000001 object_loss:4.0444 object_precision: 42.55 object_recall: 21.89 object_f1: 28.62

2022-06-20 07:45:28 - INFO - Epoch: [0] [ 800/10640] lr:0.0000001 object_loss:4.0424 object_precision: 42.62 object_recall: 21.91 object_f1: 28.66

2022-06-20 07:46:10 - INFO - Epoch: [0] [ 900/10640] lr:0.0000001 object_loss:4.0663 object_precision: 42.49 object_recall: 21.79 object_f1: 28.53
```

I use QAT followed the [guide](https://pytorch.org/docs/stable/quantization.html)
```
class MultiTaskVisionModelV4(nn.Module):
    def __init__(self, feature_extractor, feature_size, class_nums, is_quant=False):
        super(MultiTaskVisionModelV4, self).__init__()
        if is_quant:
            self.quant = torch.quantization.QuantStub()
            self.dequant = torch.quantization.DeQuantStub()
        self.is_quant = is_quant

        self.feature_extractor = feature_extractor
        self.pooling = nn.AdaptiveMaxPool2d(1)
        self.fc_object = nn.Linear(feature_size, class_nums[0])

    def forward(self, x):
        if self.is_quant:
            x = self.quant(x)

        feats = self.feature_extractor(x)
        feat = feats[-1]
        pool_op = self.pooling(feat)  # object and place

        out_op = pool_op.flatten(1)

        o1 = self.fc_object(out_op)
        if self.is_quant:
            o1 = self.dequant(o1)

        return o1


def fuse_mobilenetv2(model):
    if hasattr(model, "module"):
        model = model.module

    fuse_list = ['feature_extractor.conv_stem', 'feature_extractor.bn1', 'feature_extractor.act1']
    torch.quantization.fuse_modules(model, fuse_list, inplace=True)

    for name, module in model.feature_extractor.named_modules():
        if isinstance(module, DepthwiseSeparableConv):
            fuse_list = [['conv_dw', 'bn1', 'act1'],
                         ['conv_pw', 'bn2'],
                         ]
            torch.quantization.fuse_modules(module, fuse_list, inplace=True)
        elif isinstance(module, InvertedResidual):
            fuse_list = [['conv_pw', 'bn1', 'act1'],
                         ['conv_dw', 'bn2', 'act2'],
                         ['conv_pwl', 'bn3'],
                         ]
            torch.quantization.fuse_modules(module, fuse_list, inplace=True)
            

def convert2qat_model(model_fp32):
    model_fp32.train()
    model_fp32.qconfig = torch.quantization.get_default_qat_qconfig('qnnpack')
    torch.quantization.prepare_qat(model_fp32, inplace=True)
    return model_fp32


# modified mobilenetv2_120d
backbone = timm.create_model("mobilenetv2_120d", features_only=True, pretrained=False, out_indices=(1, 2, 3, 4))
model = MultiTaskVisionModelV4(backbone, feature_size=384, class_nums=520, is_quant=True)
fuse_mobilenetv2(model)
convert2qat_model(model)
```

I experimented to verify where the model got worse. I load a pretrain model, which trained in float32 and can get excellent result. 
I can get same output with float32 model with follows code
```
backbone = timm.create_model("mobilenetv2_120d", features_only=True, pretrained=False, out_indices=(1, 2, 3, 4))
model = MultiTaskVisionModelV4(backbone, feature_size=384, class_nums=520, is_quant=True)
model.eval()
checkpoint = torch.load("../checkpoints/pretrained_mbv2.pth", map_location="cpu")
sd = checkpoint["state_dict"]
if next(iter(sd.items()))[0].startswith('module'):
    sd = {k[len('module.'):]: v for k, v in sd.items()}
model.load_state_dict(sd)

model.train()
fuse_mobilenetv2(model)
model.eval()
```
but outputs get worse (completely wrong output) when I convert model to QAT
```
model.train()
fuse_mobilenetv2(model)
convert2qat_model(model)
model.eval()
```
strange things are coming, I removed `fuse_mobilenetv2(model)` and outputs become normal again.
I have a feeling there might be a bug here make my QAT can't converge, since fuse model should reduce the error produced by quant and dequant.

Thanks in advance

### Versions

[conda] blas                      1.0                         mkl    defaults
[conda] cpuonly                   2.0                           0    pytorch
[conda] mkl                       2020.2                      256    defaults
[conda] mkl-service               2.3.0            py36h196d8e1_0    defaults
[conda] mkl_fft                   1.3.0            py36h46781fe_0    defaults
[conda] mkl_random                1.1.1            py36h47e9c7a_0    defaults
[conda] numpy                     1.19.2           py36hadc3359_0    defaults
[conda] numpy-base                1.19.2           py36ha3acd2a_0    defaults
[conda] pytorch                   1.10.2              py3.6_cpu_0    pytorch
[conda] pytorch-mutex             1.0                         cpu    pytorch
[conda] torchaudio                0.10.2                 py36_cpu  [cpuonly]  pytorch
[conda] torchvision               0.11.3                 py36_cpu  [cpuonly]  pytorch

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo

**Comment by z-a-f** [comment]
I will investigate further. Meanwhile can you try the PTQ flow?

**Comment by JH-Xie** [comment]
> I will investigate further. Meanwhile can you try the PTQ flow?

Hi @z-a-f , nice to hear that, I'm will take a try.
Meanwhile I I have some new discoveries，I debug the code and find Conv and BN will merge to ConvBn2d defined in torch.nn.intrinsic.qat.modules.conv_fused.py, It's forward code is below.
```
    def _forward(self, input):
        assert isinstance(self.bn.running_var, torch.Tensor)
        running_std = torch.sqrt(self.bn.running_var + self.bn.eps)
        scale_factor = self.bn.weight / running_std
        weight_shape = [1] * len(self.weight.shape)
        weight_shape[0] = -1
        bias_shape = [1] * len(self.weight.shape)
        bias_shape[1] = -1

        scaled_weight = self.weight_fake_quant(self.weight * scale_factor.reshape(weight_shape))
        # using zero bias here since the bias for original conv
        # will be added later
        if self.bias is not None:
            zero_bias = torch.zeros_like(self.bias)
        else:
            zero_bias = torch.zeros(self.out_channels, device=scaled_weight.device)
        conv = self._conv_forward(input, scaled_weight, zero_bias)
        conv_orig = conv / scale_factor.reshape(bias_shape)
        if self.bias is not None:
            conv_orig = conv_orig + self.bias.reshape(bias_shape)
        conv = self.bn(conv_orig)
        return conv
```
It seems `scaled_weight` imitate merge Conv and BN operate when doing evalution, and doing quantize using merged weights. But outputs completely changed after this scale and unscale operate. The output seems right if I set scale_factor = 1.  I know that this part in tensorflow QAT is silghtly different from pytorch, since tensorflow simulate mean and std value in BN using a forward operate, but pytorch directly use source value. Hopes my discoveries can helps you.

**Comment by meixitu17** [comment]
@z-a-f @JH-Xie ,

    Do you guys have any update for this issue?

    I tried to QAT for resnet on pytorch1.09, it can't converge too.

Thanks
Jinhong

**Comment by JH-Xie** [comment]
@meixitu17 Do you use ResNet in timm? I have no trouble when using ResNet and regnet in timm, but failed to converge when I using mobilenetv2.

**Comment by meixitu17** [comment]
> @meixitu17 Do you use ResNet in timm? I have no trouble when using ResNet and RegNet in timm, but failed to converge when I using MobileNetv2.

It is my fault.  Now I found resnet can converge.
Thank you very much.

**Comment by z-a-f** [comment]
@JH-Xie Is the problem still there?

**Comment by z-a-f** [comment]
Closing this due to no movement. Feel free to reopen if the issue still presists
